core/lib/Cardano/Address/Script.hs

```diff
@@ -258,77 +321,201 @@ foldScript fn zero = \case
     foldMScripts =
         runIdentity . foldM (\acc -> Identity . foldScript fn acc) zero
 
-
 --
 -- Script validation
 --
 
 -- | Validate a 'Script', semantically
 --
 -- @since 3.0.0
-validateScript :: Script -> Either ErrValidateScript ()
-validateScript = \case
-    RequireSignatureOf (KeyHash bytes) -> do
-        when (BS.length bytes /= credentialHashSize) $ Left WrongKeyHash
+validateScript
+    :: ValidationLevel
+    -> TxValidity
+    -> Script KeyHash
+    -> Either ErrValidateScript ()
+validateScript level interval script = do
+    let validateKeyHash (KeyHash bytes) =
+            (BS.length bytes == credentialHashSize)
+    let allSigs = foldScript (:) [] script
+    unless (L.all validateKeyHash allSigs) $ Left WrongKeyHash
+    unless (requiredValidation interval script)
+        $ Left LedgerIncompatible
```

Comment:
Hmm. We loose quite a bit of information with this opaque error type. Wouldn't it be better to have `requiredValidation` return a proper error rather than a boolean? I'd suggest making a new sum-type `ErrRequiredValidation` to cover the various errors that can arise from `requiredValidation` and create a new branch in `ErrValidateScript` to hold those possible errors.

Reply:
refactored errors to reflect what is required (be compliant with ledger, not malformed in syntax sense, have legal verification keys), what is recommended. see HEAD